### ISSUE TITLE
question about the dij structure

### ISSUE THREAD

**Issue by chh105** [e0404/matRad#477]
**What is the question? Which file(s) does it concern?**
I had a question about combining dij structures from two separate forward calculations. For example, say I compute dij for a one beam plan with (couch,gantry) = (0,90) and another dij for a one beam plan with (0,-90). If I wanted to save on compute for a two beam plan with couch angles [0,0] and gantry angles [90,-90], would it be possible to combine the dij structures?

If you can combine dij structures, would I just concatenate the physical dose matrices in the column-wise direction, concatenate the other vectors (i.e. bixelNum, rayNum, beamNum) in the row-wise direction, etc?

**I have checked the following matRad Wiki pages and references:**
documentation for the dij structure

**Comment by markbangert** [comment]
In principle you can combine DIJ structures by column-wise concatenation. Thereby you kind of increase the number of bixels. Be careful that you also modify all meta information (numberOfBixels etc.) accordingly if you want to use the matrix for optimization within matRad.

However, I do not really see how this maked sense in the context of a forward calculation. There you can simply add the dose cubes?!

**Comment by chh105** [comment]
Thanks for the quick response. Also sorry if I didnt describe my use case very well, but I'm searching through a bunch of candidate angles and iteratively adding the best angle to create a n-beam plan. Most of the computation time is spent computing dij for each beam, so I was hoping to save and reuse the information

**Comment by markbangert** [comment]
In case you want to use matRad for beam angle selection you certainly should precompute the Dij matrices and reuse them. Ideally you establish some grid before and then glue them together for the combinations that you want to investigate. For such a use case, however, it might even make sense to write your own backpropagation function because the sparse matrix concatenation will also take quite some time.

Also you may want to check out approaches like https://pubmed.ncbi.nlm.nih.gov/26936695/ to save something on the optimization side.

**Comment by chh105** [comment]
Very interesting. I'll probably start by just concatenating and see if it runs fast enough. For the optimization part, I've just been running IPOPT for ~15 iterations which finishes relatively quickly (I think this is similar to the early stopping method in your paper). I've also been using an objective function with a squared deviation for the ptv and squared overdosing to the body (with equal weights). I was wondering if you tried various objective functions besides squared over/under dosing and whether there was much difference between the ones you tried.

**Comment by markbangert** [comment]
No we have not experimented with different obj functions. I have investigated different methods on the beam angle selection problem itself, though. Iterative adding, however, was pretty much clinically equivalent.

https://aapm.onlinelibrary.wiley.com/doi/abs/10.1118/1.4771932

**Comment by chh105** [comment]
Ok I'll probably just stick with the current optimization approach then. Anyways thanks for your help!

**Comment by chh105** [comment]
Just had a follow up question about combining physicalDose matrices. Say we compute two physicalDose matrices for (couch=0, gantry=90) and (couch=0, gantry=-90) and combine them by concatenating column-wise. We can also compute a physicalDose matrix for (couch=[0,0], gantry=90,-90). I would expect these two matrices to be the same (isEqual returns 1), but it seems that they are very slightly different (isEqual returns 0 and the number of nonzero terms in the physicalDose matrices differs by around 100). Is this acceptable/expected behavior?

**Comment by markbangert** [comment]
that is interesting. are the steering setups the same? meaning isocenter,
and position of beamlets, number of bixels, etc.

I would not expect them to be binary identical but right now I do not see
why they should not have the same non-zero pattern.

**Comment by chh105** [comment]
I think the pln and stf structures should be unmodified. To reproduce this manually, I setup everything in the GUI, computed three dij structures, and then compared the two physicalDose matrices (one where things are combined and one copied from the dij structure). I'm wondering if maybe I overlooked something

**Comment by chh105** [comment]
Actually sorry after checking again, I think the plan angles and isocenter are being updated by the gui, but the stf structure isn't updated.

**Comment by chh105** [comment]
I think I'm just making this more confusing with my comments, so I wrote a script to reproduce things. 
[debug_precomputed_dij.zip](https://github.com/e0404/matRad/files/5817397/debug_precomputed_dij.zip)

**Comment by markbangert** [comment]
Sorry for the late reply. Somehow your issue got covered in my mailbox. Are
you still trying to figure this out?

**Comment by chh105** [comment]
Yes I still haven't figured it out. There seems to be a small difference between the concatenated physical dose matrices and the regular physical dose matrix.

**Comment by markbangert** [comment]
Thank you for providing the script. The behaviour you observe is related to the dose influence matrix sampling. In order to save memory we perform a random sampling of the influence data in https://github.com/e0404/matRad/blob/6fad1b8c07ea4d680e5c17615c5552659897e149/matRad_calcPhotonDose.m#L279. As we are re-setting the state of the random engine only in the beginning of the entire dose calculation at https://github.com/e0404/matRad/blob/6fad1b8c07ea4d680e5c17615c5552659897e149/matRad_calcPhotonDose.m#L45 we have the funny behaviour that the first beams are binary identical while the second beam differs. If you want to make this consistent you can reset the random engine before the calculation of every beam. @wahln what do you think about including this for the main release? I am somewhat hesitant because resetting the random engine is always fishy with regard to statistical independence... But we could even think about setting the random engine based on a hash computed on the index set of the beamlets within the dij sampling function itself. That way we would even have reproducability for individual bixels given they have the exact same geometry...

**Comment by chh105** [comment]
Thanks for figuring this out so quickly. Just to clarify, are you saying that the random engine needs to be reset for each beam when computing dij for a multiple beam plan? For example, would you paste ```rng('default');``` after the for loop call in line 143 (```for i = 1:dij.numOfBeams % loop over all beams```)?

**Comment by markbangert** [comment]
yes! without having tried this out... :)

if you reset the random engine before every beam, the dij sampling pattern
should be the same for every beam as long as the stf structs are identical
- which they are in your example. this I have checked with matlab's isequal
function.

**Comment by wahln** [comment]
Nice that you guys figured that out. I think we should allow some configuration of the sampling that enforces reproducible / consistent results. I will think about some configuration parameters in pln.propDoseCalc that allow people to enable/disable & steer the sampling process / set seeds & choose seed mechanic (per beam etc.).
